main/manager.c

```diff
@@ -6873,6 +6900,203 @@ static int action_coreshowchannels(struct mansession *s, const struct message *m
 	return 0;
 }
 
+/*! \brief Destructor for channel map vector */
+static void channel_map_vector_destructor(void *obj)
+{
+	struct ast_vector_string *channel_map = obj;
+	AST_VECTOR_RESET(channel_map, ast_free);
+	AST_VECTOR_FREE(channel_map);
+}
+
+/*! \brief Helper function to add a channel name to the vector */
+static int coreshowchannelmap_add_to_map(struct ast_vector_string *v, const char *s)
+{
+	char *str;
+	int size;
+
+	/* If we've already added this channel, don't add it again */
+	if (AST_VECTOR_GET_CMP(v, s, !strcmp)) {
+                return -1;
+        }
+
+	size = strlen(s) + 1;
+	str = ast_malloc(size);
+	if (!str) {
+		ast_log(LOG_ERROR, "Unable to append channel to channel map\n");
+		return 1;
+	}
+
+	ast_copy_string(str, s, size);
+	if (AST_VECTOR_APPEND(v, str)) {
+		ast_log(LOG_ERROR, "Unable to append channel to channel map\n");
+		ast_free(str);
+		return 1;
+	}
+
+	return 0;
+}
+
+/*! \brief Recursive function to get all channels in a bridge. Follow local channels as well */
+static int coreshowchannelmap_add_connected_channels(struct ast_vector_string *channel_map,
+	struct ast_channel_snapshot *channel_snapshot, struct ast_bridge_snapshot *bridge_snapshot)
+{
+	int res = 0;
+	struct ao2_iterator iter;
+	char *current_channel_uid;
+
+	iter = ao2_iterator_init(bridge_snapshot->channels, 0);
+	while ((current_channel_uid = ao2_iterator_next(&iter))) {
+		struct ast_channel_snapshot *current_channel_snapshot;
+		int add_channel_res;
+
+		/* Don't add the original channel to the list - it's either already in there,
+		 * or it's the channel we want the map for */
+		if (!strcmp(current_channel_uid, channel_snapshot->base->uniqueid)) {
+			continue;
+		}
+
+		current_channel_snapshot = ast_channel_snapshot_get_latest(current_channel_uid);
+		if (!current_channel_snapshot) {
+			ast_log(LOG_ERROR, "Unable to get snapshot of channel\n");
+			res = 1;
+			break;
+		}
+
+		add_channel_res = coreshowchannelmap_add_to_map(channel_map, current_channel_snapshot->base->name);
+		if (add_channel_res > 0) {
+			ast_log(LOG_ERROR, "Unable to add channel to vector\n");
+			res = 1;
+			ao2_ref(current_channel_snapshot, -1);
+			break;
+		}
+
+		/* If this is a local channel that we haven't seen yet, let's go ahead and find out what else is connected to it */
+		if (!add_channel_res && ast_begins_with(current_channel_snapshot->base->name, "Local")) {
+			struct ast_channel_snapshot *other_local_snapshot;
+			struct ast_bridge_snapshot *other_bridge_snapshot;
+			char *other_local;
+			int size = strlen(current_channel_snapshot->base->name);
+
+			other_local = ast_malloc(size + 1);
```

Comment:
Same question here regarding strdup...

Reply:
I would say you don't even need to do that. Channel names are a fixed maximum size. Just use a defined buffer of that maximum size and copy the existing name in. No having to deal with memory alloc/free in that case.

Reply:
> Channel names are a fixed maximum size

I don't think that is true any longer. They are stringfields on both `struct ast_channel` and `struct ast_channel_snapshot_base` and nothing appears to enforce a maximum length.

Reply:
We still have AST_CHANNEL_NAME for the maximum size, and it's used in various places so that's... fun.